### PR TITLE
Fix RNN docstring 'Efficient implementation' example breaking on backward

### LINKED ISSUE #165621: Feedback about RNN
There is the following issue on this page: https://docs.pytorch.org/docs/stable/generated/torch.nn.RNN.html

The example code doesn't work with backward:
```python
import torch
import torch.nn as nn

# Efficient implementation equivalent to the following with bidirectional=False
rnn = nn.RNN(10, 20, 2)
params = dict(rnn.named_parameters())
def forward(x, hx=None, batch_first=False):
    if batch_first:
        x = x.transpose(0, 1)
    seq_len, batch_size, _ = x.size()
    if hx is None:
        hx = torch.zeros(rnn.num_layers, batch_size, rnn.hidden_size)
    h_t_minus_1 = hx.clone()
    h_t = hx.clone()
    output = []
    for t in range(seq_len):
        for layer in range(rnn.num_layers):
            input_t = x[t] if layer == 0 else h_t[layer - 1]
            h_t[layer] = torch.tanh(
                input_t @ params[f"weight_ih_l{layer}"].T
                + h_t_minus_1[layer] @ params[f"weight_hh_l{layer}"].T
                + params[f"bias_hh_l{layer}"]
                + params[f"bias_ih_l{layer}"]
            )
        output.append(h_t[-1].clone())
        h_t_minus_1 = h_t.clone()
    output = torch.stack(output)
    if batch_first:
        output = output.transpose(0, 1)
    return output, h_t


input = torch.randn(5, 3, 10)
h0 = torch.randn(2, 3, 20)
output, hn = forward(input, h0)

target = torch.ones_like(output)

loss = nn.CrossEntropyLoss()(output, target)
print(loss)

loss.backward()
``` 

A possible fix is:
```python
import torch
import torch.nn as nn

# Efficient implementation equivalent to the following with bidirectional=False
rnn = nn.RNN(10, 20, 2)
params = dict(rnn.named_parameters())
def forward(x, h_t_minus_1=None, batch_first=False):
    if batch_first:
        x = x.transpose(0, 1)
    seq_len, batch_size, _ = x.size()
    if h_t_minus_1 is None:
        h_t_minus_1 = torch.zeros(rnn.num_layers, batch_size, rnn.hidden_size)
    output = []
    for t in range(seq_len):
        h_t = []
        for layer in range(rnn.num_layers):
            input_t = x[t] if layer == 0 else h_t[layer - 1]
            h_t.append(torch.tanh(
                input_t @ params[f"weight_ih_l{layer}"].T
                + h_t_minus_1[layer] @ params[f"weight_hh_l{layer}"].T
                + params[f"bias_hh_l{layer}"]
                + params[f"bias_ih_l{layer}"]
            ))
        output.append(h_t[-1])
        h_t_minus_1 = h_t
    output = torch.stack(output)
    h_t = torch.stack(h_t)
    if batch_first:
        output = output.transpose(0, 1)
    return output, h_t


input = torch.randn(5, 3, 10)
h0 = torch.randn(2, 3, 20)
output, hn = forward(input, h0)

target = torch.ones_like(output)

loss = nn.CrossEntropyLoss()(output, target)
print(loss)

loss.backward()
``` 

cc @svekars @sekyondaMeta @AlannaBurke @mikaylagawarecki

### PR DESCRIPTION
**Pull request by adityasingh2400**
## Summary

Fixes #165621.

The `RNN` class docstring has an "Efficient implementation" code block intended to show users the equivalent Python loop for an `nn.RNN` forward pass. The current version assigns into `h_t[layer]` in place and rebinds `h_t_minus_1 = h_t.clone()` after writing through the same storage, so the result has no usable autograd graph and `.backward()` fails as documented in the issue.

This patch swaps the in-place writes for an append-only Python list rebuilt each timestep, matching the autograd-compatible variant proposed in the issue body. The math is unchanged: each layer still consumes the previous timestep's hidden state via `h_t_minus_1`, and the current timestep's stack of layer hidden states is exposed to the next layer via `h_t`. At the end of each timestep, `h_t` becomes `h_t_minus_1` for the next step (a plain reference rebind, no clone needed since the list of fresh tensors is never mutated).

## Test plan

- [x] Ran the new docstring example end-to-end against `torch 2.11.0`: `forward` produces output, `CrossEntropyLoss(output, target).backward()` completes, and `weight_ih_l*` / `weight_hh_l*` / `bias_*` parameters all receive non-zero gradients.
- [x] Confirmed the example block is a `.. code-block:: python` directive (not a `>>>` doctest), so it is documentation-only and not executed by xdoctest. No behavior change to `nn.RNN` itself.
- [x] `git diff` is scoped to the example inside the `RNN` class docstring in `torch/nn/modules/rnn.py`.

### CODE REVIEW

**Comment by adityasingh2400** [comment]
I just signed the EasyCLA. Could you please re-run the CLA check on this PR? It seems to be cached on the prior unsigned state.

**Comment by albanD** [comment]
Thank you very much for sending a PR.
To avoid you spending time preparing a PR for a feature we're not sure we want to add, our usual process is to discuss on the issue before doing the PR.
In particular, make sure that the issue is marked "actionable" before sending a PR!
See https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions for details!

**Comment by adityasingh2400** [comment]
Thanks albanD, and understood on the process. This one is a small documentation fix for issue #165621, where the Efficient implementation example in the RNN docstring breaks on backward because it writes in place into h_t[layer] and then rebinds h_t_minus_1 to a clone of the same storage, so the autograd graph is lost. Happy to wait for the issue to be marked actionable before you spend any review time on it. If it helps I can also trim this down to the minimal docstring correction. Just let me know how you would like to proceed.
